lib/ibm/cloud/sdk_http/base_instance.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+require_relative 'base_http_mixin'
+
+module IBM
+  module Cloud
+    # Module holds basic HTTP functionality.
+    module SDKHTTP
+      # Container that encapsulates the VPC API.
+      class BaseInstance < Hash
```

Comment:
Subclasses of Hash can be a hassle and cause issues...it's not the end of the world, but if you don't actually need all of a Hash's semantics, it might be preferable to just implement the methods you need (typically just `[]`)

Reply:
I want it to behave like a hash with all of its methods. The return could be a wrapped class or a hash, so breaking from that could be troublesome for the developer. Python has a set of safe classes to subclass Dictionary like classes, Ruby doesn't seem to have the same. Without it then the developer has to do 2 SDK calls to get to the instance class and the majority of the this hash is discarded, and only the id is used, in favour of the instance. This makes more sense from an efficiency point of view.

Reply:
I'm not sure I follow the efficiency argument with the 2 SDK calls...can you elaborate?

I'm not arguing against having a class, it's more of a composition vs inheritance discussion.  That is, _most_ callers likely only need `[]`, `fetch` and _maybe_ a known set of the `Enumerable` interface and those are trivial to write.  The downside to subclassing Hash is the unexpected things...for example:

```
BaseInstance.new.to_h.class
# => Hash
BaseInstance.new.slice(:a).refresh
# NoMethodError: undefined method `refresh' for {}:Hash
BaseInstance.new.slice(:a).class
# => Hash
```

So, many methods you do on this Hash-like thing will actually return a Hash object, as opposed to this class, which will break the expectations.  Instead, I'm suggesting if you encapsulate a Hash as an instance variable, and only delegate the methods you need/want callers to have to that ivar, it doesn't open you up to this unexpected interface.  Additionally you could expose a `.to_h` method if you want to give callers an actual Hash on which they could call anything they want.

Like I said, not a _huge_ problem, but I've personally been down this road in the past, and it has always caused unexpected behavior.

Reply:
I'm not saying it doesn't work...I'm saying it can lead to unexpected code paths because you're exposing more interface than your callers will need.

Taking your example, as a caller, what should I expect if I called `ins.size`?  You get back the number of objects in the Hash, but does that make sense from an object standpoint since it could be confused with the size of the instance?

What you want from Hash are just the fetching semantics, but by subclassing you are getting so much more that you don't need.  Adding the fetching semantics to a regular class that encapsulates the Hash is trivial and doesn't expose your object too much... For example

```ruby
 def initialize(parent, data: {}, id_key: 'id', endpoint: nil)
   # ...
   @data = data
 end

 def [](*args)
   @data[*args]
 end

 def fetch(*args)
   @data[*args]
 end

 def to_h
   @data
 end
```

or if you use the [Forwardable](https://ruby-doc.org/stdlib-2.5.1/libdoc/forwardable/rdoc/Forwardable.html) module

```ruby
 def initialize(parent, data: {}, id_key: 'id', endpoint: nil)
   # ...
   @data = data
 end

  include Forwardable
 def_delegators :@data, :[], :fetch, :to_h
```

Reply:
I get your concern. Init is super() followed by merge!(data). Which I believe should initialize the class as a hash and load the data in a standard way. It is supposed to be used as a Hash first with a couple of methods to forward to child parts of the API.

On the Instance BaseInstance class the custom methods are:
* actions -> Actions Collection with no instance. Calling all.first should return a standard hash.
initialization -> Append initialization to the instance endpoint and return the SDKResponse.
* network_interfaces -> Init & return NetworkInterfaces BaseCollection. Call all.first returns the NetworkInterface BaseInstance.
* volume_attachments -> Init & return VolumeAtttachments BaseCollection. Call all.first returns the VolumeAttachment BaseInstance.

With this use pattern there isn't a lot of in-class manipulation and a lot of querying of hashes. 

```
[2] pry(#<Sdk>)> ins = vpc.instances.all.first
=> {:bandwidth=>4000,
 :boot_volume_attachment=>
  {:device=>{:id=>"0777-de4544bd-58bb-4924-8f0c-4edf4fc994b9-8fbnl"},
   :href=>
    "https://us-....
ins = vpc.instances.all.first
[3] pry(#<Sdk>)> ls ins
IBM::Cloud::SDKHTTP::BaseHTTPMixin#methods: adhoc  delete  get  patch  post  put  unchecked_response  url
IBM::Cloud::SDKHTTP::BaseInstance#methods: connection  details  endpoint  logger  refresh  remove  token  update
IBM::Cloud::SDK::VPC::VpcHTTP#methods: metadata
IBM::Cloud::SDK::VPC::Instance#methods: actions  initialization  network_interfaces  volume_attachments
[6] pry(#<Sdk>)> vpc.instances.all.first.network_interfaces.class
=> IBM::Cloud::SDK::VPC::INSTANCE::NetworkInterfaces
[7] pry(#<Sdk>)> vpc.instances.all.first.network_interfaces.all.first.class
=> IBM::Cloud::SDK::VPC::INSTANCE::NetworkInterface
[8] pry(#<Sdk>)> vpc.instances.all.first.size
=> 20
[9] pry(#<Sdk>)> vpc.instances.all.first.key?(:bandwidth)
=> true
[10] pry(#<Sdk>)> vpc.instances.all.first.keys
=> [:bandwidth,
 :boot_volume_attachment,
 :created_at,
 :crn,
 :disks,
 :href,
 :id,
 :image,
 :memory,
 :name,
 :network_interfaces,
 :primary_network_interface,
 :profile,
 :resource_group,
 :resource_type,
 :status,
 :vcpu,
 :volume_attachments,
 :vpc,
 :zone]
[11] pry(#<Sdk>)> vpc.instances.all.first.keys.length
=> 20
```

Reply:
I can try the Forwadable module, however my worry is having to mess with self.included. I also worry that I'll just be forwarding the entirety of the Hash methods.

Reply:
The only other thing I can think of as being a gotcha is updating the underlying hash with changes fetched from a rest call. I use merge! for that but really I want to wipe the data and load a new batch.

Reply:
Add a `clear` before `merge!` in refresh. This should clear the current hash and then load the new hash.

Reply:
I have to second subclassing `Hash` being an issue but this is just temporary as we'll be moving to an OpenApi generated gem soon which has proper class definitions with accessors for all declared properties

Reply:
Next PR is using Jason's Forwardable suggestion. Wish Adam had suggested that when I asked him if there were problems with subclassing Hash.